### PR TITLE
Auto-add swap taker asset to visible asset list

### PR DESCRIPTION
**Pull request by onyb**
Also includes the following changes to `useAssetManagement` hook:
* Refactor to reduce some logic duplication.
* Remove props-drilling, inspired by https://github.com/brave/brave-core/pull/12509.
* Re-write unit-tests and improve reliability.

Resolves https://github.com/brave/brave-browser/issues/21428

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Demo

https://user-images.githubusercontent.com/3684187/160037276-2417f253-4fca-43b0-badb-7375f6994994.mov

### CODE REVIEW

**Review by Douglashdaniel — APPROVED** [review]
This is awesome! nice work!

File: components/brave_wallet_ui/common/hooks/swap.ts [hunk 1]
```diff
@@ -644,11 +644,31 @@ export default function useSwap ({ fromAsset: fromAssetProp, toAsset: toAssetPro
...
-    setToAsset(swapAssetOptions[1])
+
+    if (!fromAsset || !hasToken(fromAsset)) {
+      setFromAsset(swapAssetOptions[0])
+    }
+
+    if (!toAsset || !hasToken(toAsset)) {
+      setToAsset(swapAssetOptions[1])
+    }
   }, [swapAssetOptions])
```
Comment (josheleonard) [inline components/brave_wallet_ui/common/hooks/swap.ts:672]: FromAsset and toAsset missing from hook dependancies

**Review by josheleonard — CHANGES_REQUESTED** [review]

File: components/brave_wallet_ui/common/hooks/balance.ts [hunk 2]
```diff
@@ -21,12 +21,9 @@ export default function useBalance (networks: BraveWallet.NetworkInfo[]) {
     const tokensCoinType = getTokensCoinType(networks, token)
     // Return native asset balance
     if (
-      (
-        token.contractAddress === '' ||
-        // 0xeee... is used as a contractAddress
-        // of native assets in our Swap Widget and used for the 0x API.
-        token.contractAddress === '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
```
[the remark this answers is not in the document]
Comment (onyb) [reply components/brave_wallet_ui/common/hooks/balance.ts:28]: We still are, but here's how we're doing it now: https://github.com/brave/brave-core/pull/12745/files#diff-45697d7d6d02ec25665d4b2222d3fdaa2db9317b03c65e311d6f1fe73020ba5eR136-R137.

**Review by josheleonard — APPROVED** [review]

**Comment by darkdh** [comment]
why are we merging this PR into master when tests were all failed on CI?

**Comment by darkdh** [comment]
https://ci.brave.com/job/pr-brave-browser-f-wallet-asset-mgmt-refactor-linux/5/execution/node/226/log/
https://ci.brave.com/job/pr-brave-browser-f-wallet-asset-mgmt-refactor-macos/3/execution/node/214/log/
https://ci.brave.com/job/pr-brave-browser-f-wallet-asset-mgmt-refactor-windows/3/execution/node/233/log/

**Pull request merged**